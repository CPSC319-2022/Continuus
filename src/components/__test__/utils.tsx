import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
