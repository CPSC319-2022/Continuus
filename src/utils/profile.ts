/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
