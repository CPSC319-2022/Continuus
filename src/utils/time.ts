export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
