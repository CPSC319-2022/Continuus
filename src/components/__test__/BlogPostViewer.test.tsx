import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup } from "@testing-library/react";
import { renderWithProviders } from "./utils";

vi.mock("~/utils/time.ts", () => ({
  timeAgo: () => "X seconds ago",
}));

export const blogPosts = {
  pages: [
    {
      items: [
        {
          id: "post-1",
          userId: "user-1",
          title: "title",
          updatedAt: new Date(),
          createdAt: new Date(),
          content: "content",
          comments: [
            {
              id: "comment-1",
              userId: "user-2",
              blogPostId: "post-1",
              content: "test comment",
              createdAt: new Date(),
              updatedAt: new Date(),
            },
          ],
          user: {
            name: "test user",
            image: "imageUrl",
          },
        },
        {
          id: "post-2",
          userId: "user-3",
          title: "another title",
          updatedAt: new Date(),
          createdAt: new Date(),
          content: "another content",
          comments: [
            {
              id: "comment-2",
              userId: "user-4",
              blogPostId: "post-2",
              content: "interesting post",
              createdAt: new Date(),
              updatedAt: new Date(),
            },
          ],
          user: {
            name: "another test user",
            image: "anotherImageUrl",
          },
        },
        {
          id: "post-3",
          userId: "user-5",
          title: "different title",
          updatedAt: new Date(),
          createdAt: new Date(),
          content: "different content",
          comments: [
            {
              id: "comment-3",
              userId: "user-6",
              blogPostId: "post-3",
              content: "great information",
              createdAt: new Date(),
              updatedAt: new Date(),
            },
          ],
          user: {
            name: "different test user",
            image: "differentImageUrl",
          },
        },
      ],
    },
  ],
};

export const adminUser = {
  createdAt: new Date(),
  email: "email",
  id: "admin-id",
  image: "image",
  name: "admin",
  role: "ADMIN",
  updatedAt: new Date(),
};

const setup = (blogPosts: any = { pages: [] }) => {
  vi.mock("next-auth/react", () => ({
    useSession: () => ({
      data: {
        expires: "1",
        user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
      },
    }),
  }));

  vi.doMock("~/utils/api", () => ({
    api: {
      user: {
        currentUser: {
          useQuery: () => ({
            data: adminUser,
            isLoading: false,
          }),
        },
      },
      blogPost: {
        get: {
          useInfiniteQuery: () => ({
            data: blogPosts,
            fetchNextPage: vi.fn(),
            hasNextPage: false,
          }),
        },
      },
    },
  }));
};

describe("Blog post viewer: Snapshot", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
  });

  it("render no blog posts", async () => {
    setup();
    const { BlogPostViewer } = await import("~/components/BlogPostViewer");
    const body = renderWithProviders(
      <BlogPostViewer user="user" />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render with blog posts", async () => {
    setup(blogPosts);
    const { BlogPostViewer } = await import("~/components/BlogPostViewer");
    const body = renderWithProviders(
      <BlogPostViewer user="user" />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });
});
