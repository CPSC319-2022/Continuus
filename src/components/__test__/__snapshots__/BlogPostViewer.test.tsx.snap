// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blog post viewer: Snapshot > render no blog posts 1`] = `
<body>
  <div>
    <div
      class="w-full self-center"
    >
      Nothing to see here
    </div>
  </div>
</body>
`;

exports[`Blog post viewer: Snapshot > render with blog posts 1`] = `
<body>
  <div>
    <div
      class="w-full self-center"
    >
      <div
        class="mb-6"
      >
        <div
          class="relative"
        >
          <div
            class="card w-full rounded-md shadow-md border hover:cursor-pointer hover:border-highlight-green transition-all"
          >
            <div
              class="card-body"
            >
              <div
                class="mb-3 flex w-full justify-between"
              >
                <div
                  class="flex"
                >
                  <div
                    class="avatar self-center hover:scale-110 transition-all"
                  >
                     
                    <a
                      href="/profile/user-1"
                    >
                      <img
                        alt="avatar"
                        class="inline-block aspect-square rounded-[50%] "
                        src="imageUrl"
                        style="width: 2.5rem;"
                      />
                    </a>
                  </div>
                  <div
                    class="ml-3"
                  >
                    <a
                      href="/profile/user-1"
                    >
                      <p
                        class="text-lg font-bold hover:underline"
                      >
                        test user
                      </p>
                    </a>
                    <p
                      class="text-sm text-gray-400"
                    >
                      0 seconds ago
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="mb-3 w-fit text-2xl font-bold"
              >
                title
              </div>
              <div
                class="prose relative"
              >
                <p>
                  content
                </p>
                <div />
              </div>
              <div
                class="self-end"
              >
                <p
                  class="text-highlight-green no-underline"
                >
                  1
                   Comments
                </p>
              </div>
            </div>
          </div>
          <div
            class="absolute right-10 top-12"
          />
        </div>
      </div>
      <div
        class="mb-6"
      >
        <div
          class="relative"
        >
          <div
            class="card w-full rounded-md shadow-md border hover:cursor-pointer hover:border-highlight-green transition-all"
          >
            <div
              class="card-body"
            >
              <div
                class="mb-3 flex w-full justify-between"
              >
                <div
                  class="flex"
                >
                  <div
                    class="avatar self-center hover:scale-110 transition-all"
                  >
                     
                    <a
                      href="/profile/user-3"
                    >
                      <img
                        alt="avatar"
                        class="inline-block aspect-square rounded-[50%] "
                        src="anotherImageUrl"
                        style="width: 2.5rem;"
                      />
                    </a>
                  </div>
                  <div
                    class="ml-3"
                  >
                    <a
                      href="/profile/user-3"
                    >
                      <p
                        class="text-lg font-bold hover:underline"
                      >
                        another test user
                      </p>
                    </a>
                    <p
                      class="text-sm text-gray-400"
                    >
                      0 seconds ago
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="mb-3 w-fit text-2xl font-bold"
              >
                another title
              </div>
              <div
                class="prose relative"
              >
                <p>
                  another content
                </p>
                <div />
              </div>
              <div
                class="self-end"
              >
                <p
                  class="text-highlight-green no-underline"
                >
                  1
                   Comments
                </p>
              </div>
            </div>
          </div>
          <div
            class="absolute right-10 top-12"
          />
        </div>
      </div>
      <div
        class="mb-6"
      >
        <div
          class="relative"
        >
          <div
            class="card w-full rounded-md shadow-md border hover:cursor-pointer hover:border-highlight-green transition-all"
          >
            <div
              class="card-body"
            >
              <div
                class="mb-3 flex w-full justify-between"
              >
                <div
                  class="flex"
                >
                  <div
                    class="avatar self-center hover:scale-110 transition-all"
                  >
                     
                    <a
                      href="/profile/user-5"
                    >
                      <img
                        alt="avatar"
                        class="inline-block aspect-square rounded-[50%] "
                        src="differentImageUrl"
                        style="width: 2.5rem;"
                      />
                    </a>
                  </div>
                  <div
                    class="ml-3"
                  >
                    <a
                      href="/profile/user-5"
                    >
                      <p
                        class="text-lg font-bold hover:underline"
                      >
                        different test user
                      </p>
                    </a>
                    <p
                      class="text-sm text-gray-400"
                    >
                      0 seconds ago
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="mb-3 w-fit text-2xl font-bold"
              >
                different title
              </div>
              <div
                class="prose relative"
              >
                <p>
                  different content
                </p>
                <div />
              </div>
              <div
                class="self-end"
              >
                <p
                  class="text-highlight-green no-underline"
                >
                  1
                   Comments
                </p>
              </div>
            </div>
          </div>
          <div
            class="absolute right-10 top-12"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
